game-core/src/test/java/games/strategy/debug/ErrorMessageFormatterTest.java

```diff
@@ -0,0 +1,146 @@
+package games.strategy.debug;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+import static org.mockito.Mockito.when;
+
+import java.util.function.Function;
+import java.util.logging.LogRecord;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class ErrorMessageFormatterTest {
+
+  private static final String LOG_MESSAGE = "Pants travel with power at the stormy madagascar!";
+  private static final String EXCEPTION_MESSAGE = "Jolly, yer not trading me without a life!";
+
+  private static final Exception EXCEPTION_WITHOUT_MESSAGE = new NullPointerException();
+  private static final Exception EXCEPTION_WITH_MESSAGE = new NullPointerException(EXCEPTION_MESSAGE);
+
+  @Mock
+  private LogRecord logRecord;
+
+  private Function<LogRecord, String> errorMessageFormatter;
+
+  @BeforeEach
+  void setup() {
+    errorMessageFormatter = new ErrorMessageFormatter();
```

Comment:
This can be inlined and the field made final.

Reply:
I agree it could be. If we did then we'd be asserting/knowing that there is no state in `ErrorMessageFormatter`, which converts this test from a black box test to a white-box test. From a similar point of view, idiomatic unit testing recreates objects as part of setup so that any state between tests is removed.

White-box vs black-box testing speaks to test maintenance and becomes significant at scale. IMO that can become very important and we should in principle favor a black box test over a white box test. I do not necessarily feel that strongly about this. We could remove the interface on `ErrorMessageFormatter` and make it a final utility class and implement no API. If we want to take a preference on that, we likely should make it an explicit code convention IMO, otherwise something had to be chosen so that we can move forward and have a test.

Reply:
JUnit does create an object for each individual test run, so even if ErrorMessageFormatter changed its state this wouldn't make a difference for the test.

Reply:
Ran a simple experiment with a class that had local state and sure enough, the test class is re-instantiated between tests. I wonder if that was always the behavior of junit. The test object variable is now inlined: 4b3836ea6